arp-fingerprint.1
.\" Copyright (C) Roy Hills
.\"
.\" Copying and distribution of this file, with or without modification,
.\" are permitted in any medium without royalty provided the copyright
.\" notice and this notice are preserved.
.\"
.TH ARP-FINGERPRINT 1 "January 07, 2023"
.\" Please adjust this date whenever revising the man page.
.SH NAME
arp-fingerprint \- Fingerprint a system using ARP
.SH SYNOPSIS
.B arp-fingerprint
.RI [ options ]
.I target
.PP
The target should be specified as a single IP address or hostname.  You cannot specify multiple targets, IP networks or ranges.
.PP
If you use an IP address for the target, you can use the
.B -o
option to pass the
.B --numeric
option to
.BR arp-scan ,
which will prevent it from attempting DNS lookups.  This can speed up the
fingerprinting process, especially on systems with a slow or faulty DNS
configuration.
.SH DESCRIPTION
.B arp-fingerprint
fingerprints the specified target host using the ARP protocol.
.PP
It sends various different types of ARP request to the target, and records
which types it responds to. From this, it constructs a fingerprint string
consisting of "1" where the target responded and "0" where it did not.
An example of a fingerprint string is
.IR 01000100000 .
This fingerprint string is then used to lookup the likely target operating system.
.PP
Many of the fingerprint strings are shared by several operating systems, so
there is not always a one-to-one mapping between fingerprint strings and
operating systems. Also the fact that a system's fingerprint matches a certain
operating system (or list of operating systems) does not necessarily mean that
the system being fingerprinted is that operating system, although it is quite
likely. This is because the list of operating systems is not exhaustive; it is
just what I have discovered to date, and there are bound to be operating
systems that are not listed.
.PP
The ARP fingerprint of a system is generally a function of that system's
kernel (although it is possible for the ARP function to be implemented in
user space, it almost never is).
.PP
Sometimes, an operating system can give different fingerprints depending
on the configuration.  An example is Linux, which will respond to a non-local
source IP address if that IP is routed through the interface being tested.
This is both good and bad: on one hand it makes the fingerprinting task more
complex; but on the other, it can allow some aspects of the system configuration
to be determined.
.PP
Sometimes the fact that two different operating systems share a common ARP
fingerprint string points to a re-use of networking code. One example of
this is Windows NT and FreeBSD.
.PP
.B arp-fingerprint
uses
.B arp-scan
to send the ARP requests and receive the replies.
.PP
There are other methods that can be used to fingerprint a system using
.B arp-scan
which can be used in addition to
.BR arp-fingerprint .
These additional methods are not included in
.B arp-fingerprint
either because they are likely to cause disruption to the target system, or
because they require knowledge of the target's configuration that may not
always be available.
.PP
Most of the ARP requests that \fBarp-fingerprint\fP sends are non-standard,
so it could disrupt systems that don't have a robust TCP/IP stack.
.SH OPTIONS
.TP
.B -h
Display a brief usage message and exit.
.TP
.B -v
Display verbose progress messages.
.TP
.B -o <option-string>
Pass specified options to arp-scan. You need to enclose the options
string in quotes if it contains spaces. e.g.
-o "-I eth1".  The commonly used options are --interface (-I) and --numeric
(-N).
.TP
.B -l
Fingerprint all hosts on the local network. You do not need to specify any
target hosts if this option is given.
.TP
.B -m
Include the MAC address of the target in the output.
.SH EXAMPLES
.nf
$ arp-fingerprint 192.168.0.1
192.168.0.1   01000100000     Linux 2.2, 2.4, 2.6
.fi
.PP
.nf
$ arp-fingerprint -o "-N -I eth1" 192.168.0.202
192.168.0.202 11110100000     FreeBSD 5.3, Win98, WinME, NT4, 2000, XP, 2003
.fi
.SH NOTES
.B arp-fingerprint
is implemented in Perl, so you need to have the Perl interpreter installed on
your system to use it.
.SH "SEE ALSO"
.TP
.BR arp-scan (1)
.PP
.I https://github.com/royhills/arp-scan/wiki
The arp-scan wiki page.


NEWS
This file gives a brief overview of the major changes between each arp-scan
release.  For more details please read the ChangeLog file.

YYYY-MM-DD arp-scan 1.9.9 (in progress)

* New Features:

  - Added POSIX.1e capabilities support for systems with libcap (Linux only)
  - New --format option provides flexible output format (see manpage or help)
  - get-oui script now stores IEEE registries MA-L (OUI), MA-M, MA-S (OUI36)
    and IAB in the single file ieee-oui.txt. The ieee-iab.txt file and the
    --iabfiles option to arp-scan have been removed as they are no longer
    required.

2021-10-08 arp-scan 1.9.8 (git tag 1.9.8)

* New Features:

  - Allow the use of Linux IP aliases such as eth0:0 for the interface name.
  - Permit regular MAC addresses e.g. 00:0c:29:b9:43:1b in mac-vendor.txt.
  - --limit=n option exits after n of hosts have responded, exit 1 for <n
  - --resolve option to resolve responding IP addresses to hostnames

* Fixed bugs:

  - Potential buffer overrun in unmarshal_arp_pkt().
  - arp-scan aborts with EAGAIN on busy network or using high bandwidth
  - late ARP responses could sometimes be incorrectly flagged as duplicate

* General improvements:

  - Updated IEEE URLs in download perl scripts.
  - Updated source for Mersenne RNG and replacement strlcat/strlcpy & getopt.
  - Updated for compatability with autoconf 2.71
  - "make distcheck" works now
  - Number of responding hosts reported no longer counts duplicate packets.
  - Many typos corrected and edge cases fixed.

* Misc Changes:

  - CI framework migrated from travis-ci to github actions.
  - Several new tests for "make check"

2019-11-10 arp-scan 1.9.7

* Improved error messages from libpcap functions.

* Remove obsolescent and unused autoconf macros.  arp-scan 1.9.7 assumes that
  the C compiler is ANSI C (C89) compliant.

2019-10-13 arp-scan 1.9.6

* Use libpcap function pcap_set_immediate_mode() instead of ioctl calls to
  ensure packets are delivered immediately. This fixes the bug which caused
  arp-scan on linux to not report any hosts with libpcap 1.9.1. This change
  means arp-scan now requires libpcap 1.5.0 or later and will not work with
  earlier versions.

* Fix compiler warnings caused by the depreciated function pcap_lookupdev()
  in libpcap 1.9.0 and later.

2016-09-03 arp-scan 1.9.5

* Use posix hash table functions hcreate(), hsearch() and hdestroy() instead
  of the gas hash table code. Thanks to nihilus for the suggestion.

* Remove function replacement for inet_aton, as this was only required for
  Solaris 8, which is now considered obsolete.

* Added arp-fingerprint patterns for FreeBSD 10.3, DragonflyBSD 4.6, Windows10,
  Linux 4.0, Linux 4.6, OpenBSD 5.9, NetBSD 7.0.

* Added "-l" option to arp-fingerprint to support fingerprinting all hosts on
  the local network. Thanks to Rhig for the pull request.

* Use the source_mac rather than interface_mac in the pcap filter, to permit
  reception of packets with spoofed MAC source address. Thanks to tissieres
  for the pull request.

* Use the libpcap 1.0 API functions pcap_create() instead of pcap_open_live().
  This means that arp-scan now requires libpcap 1.0 or later and will not work
  with earlier libpcap versions.

* Updated IEEE OUI and IAB download locations to reflect IEEE website changes.

* Updated IEEE OUI and IAB MAC/Vendor files.  There are now 22487 OUI entries
  and 4575 IAB entries.

2013-11-24 arp-scan 1.9.2

* Added new --plain (-x) option to suppress printing of header and footer text,
  and only display one output line for each responding host. Idea from Stefan
  Tomanek's arp-scan fork on github at https://github.com/wertarbyte/arp-scan.

* Use LWP::UserAgent instead of LWP::Simple in get-oui and get-iab to enable
  the raw content to be obtained, which avoids Unicode/UTF-8 issues.

* Added arp-fingerprint patterns for WIZnet W5100 and Cisco IOS 15.0.

* Moved arp-scan development from internal SVN repository to github at
  https://github.com/royhills/arp-scan. The move to git means that the commit
  object names are now SHA1 hashes instead of increasing integer values. So
  they are longer suitable for internal file versions with rcsid variables.
  Accordingly the rcsid variables have been removed.

2013-07-24 arp-scan 1.9:

* Updated IEEE OUI and IAB MAC/Vendor files.  There are now 18157 OUI entries
  and 4414 IAB entries.

* Use autoconf 2.69 and automake 1.11 to add support for ARM 64-bit CPUs.

* Use libpcap functions to obtain the interface IP address and send the ARP
  packet, instead of using our own link-layer specific functions. The only
  link-layer specific function that we still need is get_hardware_address()
  to obtain the interface MAC address. This means we now require libpcap 0.9.3
  or later.

* Added support for Dragonfly BSD.

* The -u option to get-iab and get-oui scripts now works.

* get-oui and get-iab scripts now get the OUI and IAB files from the new
  locations on the IEEE website, and allow whitespace at the beginning of
  the line.

* If the MAC/Vendor file locations are not explicitly specified, look for them
  in the current directory and then in their default location.

* Raised default timeout from 100ms to 500ms.

* Added new --rtt (-D) option to display the packet round-trip time.

* Include <net/bpf.h> header file early in link-bpf.c to avoid BPF symbol
  problems on some BSD based operating systems.

* Added arp-fingerprint patterns for GNU/Hurd, Amazon Kindle (Linux 2.6),
  BeOS, Windows 8, Recent Linux, FreeBSD, NetBSD and OpenBSD versions, and
  RiscOS.

* Added data file "pkt-custom-request-vlan-llc.dat" to the tarball to allow
  the ARP request packet generation self test to complete successfully.

* Various minor bug fixes and improvements.

2011-03-01 arp-scan 1.8:

* Updated IEEE OUI and IAB MAC/Vendor files.  There are now 14707 OUI entries
  and 3542 IAB entries.

* Added support for trailer ARP replies, which were used in early versions
  of BSD Unix on VAX.


arp-fingerprint.pl
#!/usr/bin/perl
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
#
# $Id$
# arp-fingerprint.pl -- Perl script to fingerprint system with arp-scan
#
# Author: Roy Hills
# Date: 30th May 2006
#
# This script uses arp-scan to fingerprint the operating system on the
# specified target.
#
# It sends various different ARP packets to the target, and records which
# ones it responds to.  From this, it constructs a fingerprint string
# which is used to match against a hash containing known fingerprints.
#
use warnings;
use strict;
use Getopt::Std;
#
# You might need to change this to suit your configuration
# You can change the interface by defining RMIF, e.g.
# RMIF=eth1 arp-fingerprint.pl 192.168.0.1
my $arpscan="arp-scan -N -q -r 1";
#
# Hash of known fingerprints
#
# These fingerprints were observed on:
#
# FreeBSD 5.3	FreeBSD 5.3 on VMware
# FreeBSD 4.3	FreeBSD 4.3 on VMware
# Win98		Windows 98 SE on VMware
# NT4		Windows NT Workstation 4.0 SP6a on Pentium
# 2000		Windows 2000
# XP		Windows XP Professional SP2 on Intel P4
# 2003		Windows 2003 Server SP1 on Intel P4
# Linux 2.0	Linux 2.0.29 on VMware (debian 1.3.1)
# Linux 2.2	Linux 2.2.19 on VMware (debian potato)
# Linux 2.4	Linux 2.4.29 on Intel P3 (debian sarge)
# Linux 2.6	Linux 2.6.15.7 on Intel P3 (debian sarge)
# Cisco IOS	IOS 12.0(8) on Cisco 1601, IOS 12.1(27b) on Cisco 2621, IOS 12.2(32) on Cisco 1603, IOS 12.3(15) on Cisco 2503
# Solaris 2.6	Solaris 2.6 (SPARC) on Sun Ultra 5
# Solaris 7	Solaris 7 (x86) on VMware
# Solaris 8	Solaris 8 (SPARC) on Sun Ultra 5 (64 bit)
# Solaris 9	Solaris 9 (SPARC) on Sun Ultra 5 (64 bit)
# Solaris 10	Solaris 10 (x86) on VMware
# ScreenOS 5.0	ScreenOS 5.0.0r9 on NetScreen 5XP
# MacOS 10.4	MacOS 10.4.6 on powerbook G4
# MacOS 10.3	MacOS 10.3.9 on imac G3
# IRIX 6.5	IRIX64 IRIS 6.5 05190004 IP30 on SGI Octane
# SCO OS 5.0.7	SCO OpenServer 5.0.7 on VMware
# Win 3.11	Windows for Workgroups 3.11/DOS 6.22 on VMware
# 95		Windows 95 OSR2 on VMware
# NT 3.51	Windows NT Server 3.51 SP0 on VMware
# 4.3 BSD	4.3 BSD (Quasijarus0c) on MicroVAX 3000 (SIMH simulated)
# OpenBSD 3.1	OpenBSD 3.1 on VMware
# OpenBSD 3.9	OpenBSD 3.9 on VMware
# NetBSD 2.0.2	NetBSD 2.0.2 on VMware
# IPSO 3.2.1	IPSO 3.2.1-fcs1 on Nokia VPN 210
# Netware 6.5	Novell NetWare 6.5 on VMware
# PIX OS	PIX OS on PIX Firewall at 7663
# WinME		Windows ME on VMware
#
my %fp_hash = (
   '11110100000' => 'FreeBSD 5.3, Win98, WinME, NT4, 2000, XP, 2003',
   '01000100000' => 'Linux 2.2, 2.4, 2.6',
   '01010100000' => 'Linux 2.2, 2.4, 2.6',	# If non-local IP is routed
   '00000100000' => 'Cisco IOS 12.0, 12.1, 12.2, 12.3',
   '11110110000' => 'Solaris 2.6, 7, 8, 9, 10',
   '01000111111' => 'ScreenOS 5.0',
   '11110000000' => 'Linux 2.0, MacOS 10.4, IPSO 3.2.1',
   '11110100011' => 'MacOS 10.3, FreeBSD 4.3, IRIX 6.5',
   '10010100011' => 'SCO OS 5.0.7',
   '10110100000' => 'Win 3.11, 95, NT 3.51',
   '11110000011' => '4.3 BSD, OpenBSD 3.1, OpenBSD 3.9',
   '10110110000' => 'NetBSD 2.0.2',
   '00010110011' => 'Unknown 1', # dwk at 7663 in June 2006, Entrada Networks
   '01010110011' => 'PIX OS', # dwk at 7663 in June 2006, Cisco
   '00000110000' => 'Netware 6.5',
   );
#
my $usage =
qq/Usage: arp-fingerprint.pl [options] <target>
Fingerprint the target system using arp-scan.

'options' is one or more of:
        -h Display this usage message.
        -v Give verbose progress messages.
/;
my %opts;
my $verbose;
my $fingerprint="";
my $fp_name;
#
# Process options
#
die "$usage\n" unless getopts('hv',\%opts);
if ($opts{h}) {
   print "$usage\n";
   exit(0);
}
$verbose=$opts{v} ? 1 : 0;
#
if ($#ARGV != 0) {
   die "$usage\n";
}
my $target=shift;
#
# Check that the system responds to an arp-scan with no options.
# If it does, then fingerprint the target.
#
if (&fp("","$target") eq "1") {
# 1: source protocol address = localhost
   $fingerprint .= &fp("--arpspa=127.0.0.1","$target");
# 2: source protocol address = zero
   $fingerprint .= &fp("--arpspa=0.0.0.0","$target");
# 3: source protocol address = broadcast
   $fingerprint .= &fp("--arpspa=255.255.255.255","$target");
# 4: source protocol address = non local (network 1 is reserved)
   $fingerprint .= &fp("--arpspa=1.0.0.1","$target");	# Non-local source IP
# 5: invalid arp opcode
   $fingerprint .= &fp("--arpop=255","$target");
# 6: arp hardware type = IEEE_802.2
   $fingerprint .= &fp("--arphrd=6","$target");
# 7: invalid arp hardware type
   $fingerprint .= &fp("--arphrd=255","$target");
# 8: invalid arp protocol type
   $fingerprint .= &fp("--arppro=0xffff","$target");
# 9: arp prototocol type = Novell IPX
   $fingerprint .= &fp("--arppro=0x8137","$target");
# 10: invalid protocol address length
   $fingerprint .= &fp("--arppln=6","$target");
# 11: Invalid hardware address length
   $fingerprint .= &fp("--arphln=8","$target");
#
   if (defined $fp_hash{$fingerprint}) {
      $fp_name = "$fp_hash{$fingerprint}";
   } else {
      $fp_name = "UNKNOWN";
   }
   print "$target\t$fingerprint\t$fp_name\n";
} else {
   print "$target\tNo Response\n";
}
#
# Scan the specified IP address with arp-scan using the given options.
# Return "1" if the target responds, or "0" if it does not respond.
#
sub fp ($$) {
   my $ip;
   my $options;
   my $response = "0";
   ($options, $ip) = @_;

   open(ARPSCAN, "$arpscan $options $ip |") || die "arp-scan failed";
   while (<ARPSCAN>) {
      if (/^[0-9]+\.[0-9]+\.[0-9]+\.[0-9]+\t/) {
         $response = "1";
         last;
      }
   }
   close(ARPSCAN);

   if ($verbose) {
      if ($response) {
         print "$options\tYes\n";
      } else {
         print "$options\tNo\n";
      }
   }

   return $response;
}
